File: du-backend/src/main/java/com/dwin/du/api/service/ReceiptService.java
package com.dwin.du.api.service;
import com.dwin.du.api.dto.ReceiptDto;
import com.dwin.du.api.entity.*;
import com.dwin.du.api.repository.ReceiptRepository;
import com.dwin.du.api.request.AddEditReceiptRequest;
import com.dwin.du.api.repository.PersonProductRepository;
import com.dwin.du.api.repository.ProductRepository;
import com.dwin.du.update.EntityUpdateService;
import com.dwin.du.validation.ValidationException;
import com.dwin.du.validation.ValidationService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ReceiptService {

    private final ReceiptRepository receiptRepository;
    private final ProductRepository productRepository;
    private final PersonProductRepository personProductRepository;
    private final EntityUpdateService updater;
    private final ValidationService validator;

    public ResponseEntity<?> addReceipt(String username, AddEditReceiptRequest request) {
        try {
            User user = validator.validateUser(username);
            validator.isNull(request, "Nie przekazano danych");
            validator.isEmpty(request.getName(), "Nazwa rachunku jest wymagana");

            Receipt receipt = Receipt.builder()
                    .user(user)
                    .name(request.getName())
                    .date(request.getDate())
                    .totalPrice(0.0)
                    .settled(false)
                    .build();

            receiptRepository.save(receipt);
            updater.updatePerson(username);
            return ResponseEntity.ok().build();

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> editReceipt(String username, int receiptId, AddEditReceiptRequest request) {
        try {
            validator.validateUser(username);
            validator.isNull(request, "Nie przekazano danych");
            validator.isEmpty(request.getName(), "Nazwa rachunku jest wymagana");
            Receipt receipt = validator.validateReceipt(username, receiptId);

            receipt.setName(request.getName());
            receipt.setDate(request.getDate());

            receiptRepository.save(receipt);
            return ResponseEntity.ok().build();

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> removeReceipt(String username, int receiptId) {
        try {
            validator.validateUser(username);
            validator.isNull(receiptId, "Brak identyfikatora rachunku");
            Receipt receipt = validator.validateReceipt(username, receiptId);

            List<Product> products = productRepository.findByReceipt(receipt);

            Set<Person> involvedPersons = new HashSet<>();
            for (Product product : products) {
                List<PersonProduct> personProducts = personProductRepository.findByProduct(product);
                involvedPersons.addAll(
                        personProducts.stream().map(PersonProduct::getPerson).collect(Collectors.toSet())
                );
                personProductRepository.deleteAll(personProducts);
            }

            productRepository.deleteAll(products);
            receiptRepository.delete(receipt);
            updater.updatePerson(username);
            return ResponseEntity.ok().build();

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> setTotalPrice(String username, int receiptId, Double totalPrice) {
        try {
            validator.validateUser(username);
            validator.isNull(receiptId, "Brak identyfikatora rachunku");
            validator.isNull(totalPrice, "Kwota łączna jest wymagana");
            Receipt receipt = validator.validateReceipt(username, receiptId);

            receipt.setTotalPrice(totalPrice);

            receiptRepository.save(receipt);
            updater.updatePerson(username);
            return ResponseEntity.ok().build();

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> setIsSettled(String username, int receiptId, boolean settled) {
        try {
            validator.validateUser(username);
            validator.isNull(receiptId, "Brak identyfikatora rachunku");
            Receipt receipt = validator.validateReceipt(username, receiptId);

            receipt.setSettled(settled);
            receiptRepository.save(receipt);

            List<Product> products = productRepository.findByReceipt(receipt);
            for (Product product : products) {
                product.setSettled(settled);
                productRepository.save(product);

                List<PersonProduct> personProducts = personProductRepository.findByProduct(product);
                for (PersonProduct personProduct : personProducts) {
                    personProduct.setSettled(settled);
                    personProductRepository.save(personProduct);
                }
            }

            updater.updatePerson(username);
            return ResponseEntity.ok().build();

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getReceipt(String username, int receiptId) {
        try {
            validator.validateUser(username);
            validator.isNull(receiptId, "Brak identyfikatora rachunku");
            Receipt receipt = validator.validateReceipt(username, receiptId);

            Receipt response = Receipt.builder()
                    .id(receipt.getId())
                    .name(receipt.getName())
                    .date(receipt.getDate())
                    .totalPrice(receipt.getTotalPrice())
                    .settled(receipt.isSettled())
                    .user(receipt.getUser())
                    .build();

            return ResponseEntity.ok(response);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getReceipts(String username) {
        try {
            User user = validator.validateUser(username);

            List<Receipt> receipts = receiptRepository.findByUser(user);
            List<ReceiptDto> responseList = new ArrayList<>();
            for (Receipt receipt : receipts) {
                ReceiptDto response = ReceiptDto.builder()
                        .id(receipt.getId())
                        .name(receipt.getName())
                        .date(receipt.getDate())
                        .totalPrice(receipt.getTotalPrice())
                        .isSettled(receipt.isSettled())
                        .build();
                responseList.add(response);
            }

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getReceiptsByDataRange(String username, String fromDate, String toDate) throws ValidationException {
        try {
            User user = validator.validateUser(username);
            validator.isEmpty(fromDate, "Data początkowa jest wymagana");
            validator.isEmpty(toDate, "Data końcowa jest wymagana");

            SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy");
            Date dateFrom = dateFormat.parse(fromDate);
            Date dateTo = dateFormat.parse(toDate);

            dateFrom = new Date(dateFrom.getTime());
            dateFrom.setHours(0);
            dateFrom.setMinutes(0);
            dateFrom.setSeconds(0);

            dateTo = new Date(dateTo.getTime());
            dateTo.setHours(23);
            dateTo.setMinutes(59);
            dateTo.setSeconds(59);

            List<Receipt> receipts = receiptRepository.findByUserAndDateBetween(user, dateFrom, dateTo);
            List<ReceiptDto> responseList = receipts.stream()
                    .map(receipt -> ReceiptDto.builder()
                            .id(receipt.getId())
                            .name(receipt.getName())
                            .date(receipt.getDate())
                            .totalPrice(receipt.getTotalPrice())
                            .isSettled(receipt.isSettled())
                            .build())
                    .collect(Collectors.toList());

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }


    private ResponseEntity<?> handleException(Exception e) {
        if (e instanceof ValidationException) {
            HttpStatus status = ((ValidationException) e).getStatus();
            return ResponseEntity.status(status).body(e.getMessage());
        }
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Wystąpił nieoczekiwany błąd.");
    }
}

File: divvyup/DivvyUp.Web/Service/ReceiptService.cs
using System.Globalization;
using System.Net;
using AutoMapper;
using DivvyUp.Web.Data;
using DivvyUp.Web.Update;
using DivvyUp.Web.Validation;
using DivvyUp_Impl_Maui.Api.Exceptions;
using DivvyUp_Shared.Dto;
using DivvyUp_Shared.Interface;
using DivvyUp_Shared.Model;
using DivvyUp_Shared.RequestDto;
using Microsoft.EntityFrameworkCore;


namespace DivvyUp.Web.Service
{
    public class ReceiptService : IReceiptService
    {
        private readonly MyDbContext _dbContext;
        private readonly IMapper _mapper;
        private readonly MyValidator _validator;
        private readonly EntityUpdateService _entityUpdateService;
        private readonly UserContext _userContext;

        public ReceiptService(
            UserContext userContext,
            MyDbContext dbContext,
            IMapper mapper,
            MyValidator validator,
            EntityUpdateService entityUpdateService)
        {
            _userContext = userContext;
            _dbContext = dbContext;
            _mapper = mapper;
            _validator = validator;
            _entityUpdateService = entityUpdateService;
        }

        public async Task Add(AddEditReceiptRequest request)
        {
            _validator.IsNull(request, "Nie przekazano danych");
            _validator.IsNull(request.Date, "Data jet wymagana");
            _validator.IsEmpty(request.Name, "Nazwa jest wymagana");
            var user = await _userContext.GetCurrentUser();

            var newReceipt = new Receipt()
            {
                User = user,
                Name = request.Name,
                Date = request.Date,
                TotalPrice = 0,
                Settled = false,
            };

            _dbContext.Receipts.Add(newReceipt);
            await _dbContext.SaveChangesAsync();
        }

        public async Task Edit(AddEditReceiptRequest request, int receiptId)
        {
            _validator.IsNull(request, "Nie przekazano danych");
            _validator.IsNull(request.Date, "Data jet wymagana");
            _validator.IsEmpty(request.Name, "Nazwa jest wymagana");
            _validator.IsNull(receiptId, "Brak identyfikatora rachunku");
            var user = await _userContext.GetCurrentUser();
            var receipt = await _validator.GetReceipt(user, receiptId);

            receipt.Name = request.Name;
            receipt.Date = request.Date;

            _dbContext.Receipts.Update(receipt);

            await _dbContext.SaveChangesAsync();
        }

        public async Task Remove(int receiptId)
        {
            _validator.IsNull(receiptId, "Brak identyfikatora rachunku");
            var user = await _userContext.GetCurrentUser();
            var receipt = await _validator.GetReceipt(user, receiptId);

            var products = await _dbContext.Products
                .Where(p => p.ReceiptId == receiptId)
                .ToListAsync();

            var personProductsToRemove = new List<PersonProduct>();

            foreach (var product in products)
            {
                var personProducts = await _dbContext.PersonProducts
                    .Where(pp => pp.ProductId == product.Id)
                    .ToListAsync();

                personProductsToRemove.AddRange(personProducts);
            }

            _dbContext.PersonProducts.RemoveRange(personProductsToRemove);
            _dbContext.Products.RemoveRange(products);
            _dbContext.Receipts.Remove(receipt);

            await _dbContext.SaveChangesAsync();
            await _entityUpdateService.UpdatePerson(user, false);
        }

        public async Task SetSettled(int receiptId, bool settled)
        {
            _validator.IsNull(receiptId, "Brak identyfikatora rachunku");
            var user = await _userContext.GetCurrentUser();
            var receipt = await _validator.GetReceipt(user, receiptId);

            receipt.Settled = settled;

            var products = await _dbContext.Products.Where(p => p.ReceiptId == receiptId).ToListAsync();
            foreach (var product in products)
            {
                product.Settled = settled;

                var personProducts = await _dbContext.PersonProducts
                    .Where(pp => pp.ProductId == product.Id)
                    .ToListAsync();

                foreach (var personProduct in personProducts)
                {
                    personProduct.Settled = settled;
                }

                _dbContext.PersonProducts.UpdateRange(personProducts);
            }

            _dbContext.Receipts.Update(receipt);
            _dbContext.Products.UpdateRange(products);
            await _dbContext.SaveChangesAsync();
            await _entityUpdateService.UpdatePerson(user, false);
        }

        public async Task<ReceiptDto> GetReceipt(int receiptId)
        {
            _validator.IsNull(receiptId, "Brak identyfikatora rachunku");
            var user = await _userContext.GetCurrentUser();
            var receipt = await _validator.GetReceipt(user, receiptId);
            var receiptDto = _mapper.Map<ReceiptDto>(receipt);
            return receiptDto;
        }

        public async Task<List<ReceiptDto>> GetReceipts()
        {
            var user = await _userContext.GetCurrentUser();
            var receipts = await _dbContext.Receipts
                .AsNoTracking()
                .Where(p => p.UserId == user.Id)
                .ToListAsync();

            var receiptsDto = _mapper.Map<List<ReceiptDto>>(receipts).ToList();
            return receiptsDto;
        }

        public async Task<List<ReceiptDto>> GetReceiptsByDataRange(DateOnly from, DateOnly to)
        {
            _validator.IsNull(from, "Brak daty od");
            _validator.IsNull(to, "Brak daty do");
            var user = await _userContext.GetCurrentUser();

            if (from > to)
            {
                throw new DException(HttpStatusCode.BadRequest, "Data od nie może być późniejsza niż data do");
            }

            var receipts = await _dbContext.Receipts
                .AsNoTracking()
                .Where(p => p.UserId == user.Id && p.Date >= from && p.Date <= to)
                .ToListAsync();

            var receiptsDto = _mapper.Map<List<ReceiptDto>>(receipts);
            return receiptsDto;
        }

    }
}
